File: 0078-subsets/0078-subsets.java
import java.util.ArrayList;
import java.util.List;

class Solution {
    public void generate(List<Integer> arr, int idx, List<Integer> subset, List<List<Integer>> result) {
        if (idx == arr.size()) {
            result.add(new ArrayList<>(subset));
            return;
        }
        generate(arr, idx + 1, subset, result);
        subset.add(arr.get(idx));
        generate(arr, idx + 1, subset, result);
        subset.remove(subset.size() - 1);
    }

    public List<List<Integer>> subsets(int[] arr) {
        List<Integer> inputList = new ArrayList<>();
        for (int num : arr) {
            inputList.add(num);
        }

        List<Integer> subset = new ArrayList<>();
        List<List<Integer>> result = new ArrayList<>();
        generate(inputList, 0, subset, result);
        return result;
    }
}

File: 0078-subsets/0078-subsets.py
class Solution(object):
    def subsets(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        n = len(nums)
        def rec(i,c):
            if i < 0:
                ans.append(list(c))
                return
            rec(i-1,c+[nums[i]])
            rec(i-1,c)
        ans = list()
        rec(n-1,[])
        return ans

        